### PR TITLE
add tests to assert on expected behavior of  wrapped ResponseWriter objects

### PR DESCRIPTION
**Pull request by tkashem**
#### What type of PR is this?

/kind cleanup

#### What this PR does / why we need it:
Add tests to verify expected behavior of the `ResponseWriter` object(s) wrapped by the handlers in the handler chain

#### Which issue(s) this PR fixes:
Fixes #

#### Does this PR introduce a user-facing change?
```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs
- [KEP]: https://github.com/kubernetes/enhancements/pull/4462
```

### CODE REVIEW

File: staging/src/k8s.io/apiserver/pkg/endpoints/filters/webhook_duration_test.go [hunk 1]
```diff
@@ -0,0 +1,54 @@
+/*
+Copyright 2023 The Kubernetes Authors.
```
Comment (sttts) [inline staging/src/k8s.io/apiserver/pkg/endpoints/filters/webhook_duration_test.go:2]: 2024

File: staging/src/k8s.io/apiserver/pkg/endpoints/filters/webhook_duration_test.go [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+	// FakeResponseWriter does not implement http.Flusher, FlusherError,
+	// http.CloseNotifier, or http.Hijacker; so WrapForHTTP1Or2 is not
+	// expected to return an outer object.
+	if outer != middle {
+		t.Errorf("did not expect a new outer object, but got %v", outer)
+	}
+
+	decorator, ok := outer.(responsewriter.UserProvidedDecorator)
+	if !ok {
+		t.Errorf("expected the middle to implement UserProvidedDecorator")
```
Comment (sttts) [inline staging/src/k8s.io/apiserver/pkg/endpoints/filters/webhook_duration_test.go:40]: t.Fatal

File: staging/src/k8s.io/apiserver/pkg/endpoints/filters/webhook_duration_test.go [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+func TestLatencyTrackerResponseWriterConstruction(t *testing.T) {
+	inner := &responsewritertesting.FakeResponseWriter{}
+	middle := &writeLatencyTracker{ResponseWriter: inner}
+	outer := responsewriter.WrapForHTTP1Or2(middle)
+
+	// FakeResponseWriter does not implement http.Flusher, FlusherError,
+	// http.CloseNotifier, or http.Hijacker; so WrapForHTTP1Or2 is not
+	// expected to return an outer object.
+	if outer != middle {
+		t.Errorf("did not expect a new outer object, but got %v", outer)
```
Comment (sttts) [inline staging/src/k8s.io/apiserver/pkg/endpoints/filters/webhook_duration_test.go:35]: t.Fatal?

Comment (tkashem) [reply staging/src/k8s.io/apiserver/pkg/endpoints/filters/webhook_duration_test.go:35]: the test can still continue to the next assertion, so using `Errorf` here, thoughts?

Comment (sttts) [reply staging/src/k8s.io/apiserver/pkg/endpoints/filters/webhook_duration_test.go:35]: oh, then it's fine. Wasn't sure about that.

File: staging/src/k8s.io/apiserver/pkg/endpoints/responsewriter/testing/fake.go [hunk 4]
```diff
@@ -89,3 +89,26 @@ func AssertResponseWriterInterfaceCompatibility(t *testing.T, inner, outer http.
 		t.Errorf("Expected the inner and outer http.ResponseWriter object to be compatible with http.Hijacker, but got - inner: %t, outer: %t", innerHijackable, outerHijackable)
 	}
 }
+
+func AssertResponseWriter(t *testing.T, w http.ResponseWriter, req *http.Request) {
```
Comment (sttts) [inline staging/src/k8s.io/apiserver/pkg/endpoints/responsewriter/testing/fake.go:93]: assert what?

File: staging/src/k8s.io/apiserver/pkg/endpoints/responsewriter/wrapper_test.go [hunk 5]
```diff
@@ -29,108 +29,68 @@ import (
 	responsewritertesting "k8s.io/apiserver/pkg/endpoints/responsewriter/testing"
 )
```
Comment (tkashem) [inline staging/src/k8s.io/apiserver/pkg/endpoints/responsewriter/wrapper_test.go:32]: this commit does not introduce any change in the test, it just merges the two tests into one.

**Comment by tkashem** [comment]
sharing stress test results here:
```
$ go test -race k8s.io/apiserver/pkg/server -c
$ stress -p=32 ./server.test -test.run=TestWrappedResponseWriterWithFullHandlerChain
5s: 76 runs so far, 0 failures
10s: 188 runs so far, 0 failures
15s: 304 runs so far, 0 failures
...
7m50s: 8499 runs so far, 0 failures
7m55s: 8590 runs so far, 0 failures
8m0s: 8682 runs so far, 0 failures
```

File: staging/src/k8s.io/apiserver/pkg/endpoints/responsewriter/testing/fake.go (hunk 4 above)
Comment (tkashem) [reply staging/src/k8s.io/apiserver/pkg/endpoints/responsewriter/testing/fake.go]: renamed to `AssertResponseWriterImplementsExtendedInterfaces`

**Comment by tkashem** [comment]
@sttts @wojtek-t I have split the commits to make review easier, and shared the `stress` test result above

**Pull request closed without merging**